File: javascript/exercises.js
// Change Function
export function change(amount) {
  if (!Number.isInteger(amount)) {
    throw new TypeError("Amount must be an integer")
  }
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative")
  }
  let [counts, remaining] = [{}, amount]
  for (const denomination of [25, 10, 5, 1]) {
    counts[denomination] = Math.floor(remaining / denomination)
    remaining %= denomination
  }
  return counts
}

// First then LowerCase 
export function firstThenLowerCase(word, predicate) {
  return word.find(predicate)?.toLowerCase();
}

// Powers Generator
export function* powersGenerator({ofBase, upTo}) {
  let power = 0
  let value = ofBase ** power
  while (value <= upTo) {
    yield value
    power += 1
    value = ofBase ** power
  }
}


// Write your say function here
export function say(word) {
  if (word === undefined){
    return ''
  }
  let sentence = word;
  function chain(nextWord) {
     
      if (nextWord !== undefined) {
          sentence += (nextWord === "" ? " " : " " + nextWord);
          return chain; // Allow chaining
      }
      return sentence
  }
  return chain;
}


// Line Count
export async function meaningfulLineCount(filename) {
  let count = 0
  const file = await open(filename, "r")
  for await (const line of file.readLines()) {
    // Note that readLines will autoclose the file, yay
    const trimmed = line.trim()
    if (trimmed && !trimmed.startsWith("#")) {
      count++
    }
  }
  return count
}


// Quaternion Class
export class Quaternion {
  constructor(a,b, c,d) {
    
    Object.assign(this, {a,b,c,d })
    
    Object.freeze(this)
  }

  plus(v) {
    return new Quaternion(this.a + v.a, this.b + v.b, this.c + v.c, this.d + v.d)
  }
  times(v) {
    //equation from stack overflow https://stackoverflow.com/questions/19956555/how-to-multiply-two-quaternions
    const product =[
    this.a * v.a - this.b * v.b - this.c * v.c - this.d * v.d,  
    this.a * v.b + this.b * v.a + this.c * v.d - this.d * v.c,  
    this.a * v.c - this.b * v.d + this.c * v.a + this.d * v.b,  
    this.a * v.d + this.b * v.c - this.c * v.b + this.d * v.a   

    ]
    return new Quaternion(product[0],product[1],product[2],product[3])
  }
  equals(v) {
    return this.a === v.a && this.b === v.b && this.c === v.c && this.d === v.d
  }
  get conjugate() {
    return new Quaternion(this.a, -this.b, -this.c, -this.d)
  }
  get coefficients() {
    
    return [this.a,this.b,this.c,this.d]
  }

  toString() {
  let result = '';
  const symbols = ['', 'i', 'j', 'k']; 
  const components = [this.a, this.b, this.c, this.d];
  
  // Iterate through each component
  components.forEach((value, idx) => {
    if (value !== 0) {
     
            if (value > 0 && result) {
        result += '+';
      } else if (value < 0) {
        result += '-';
      }

      const absValue = Math.abs(value);
      if (absValue !== 1 || idx === 0) {
        result += absValue;
      }

      result += symbols[idx];
    }
  });
  return result || '0'; 
}
}

File: python/exercises.py
from dataclasses import dataclass
from collections.abc import Callable


#Change function
def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts


# First then lower case
def first_then_lower_case( word: list[str], predicate: Callable, /) -> str:
    for string in word:
        if predicate(string):
            return string.lower()
        

# Powers Generator
def powers_generator(*,base: int, limit: int):
    power = 0
    result = 1
    while result <= limit:
        yield result
        power += 1
        result = base ** power


# Say Function
def say(word=None, /):
    if word == None:
        return ""
    def chain(next_word=None):
        if next_word == None:
            return word
        return say(f"{word} {next_word}")
    return chain

# Line count
def meaningful_line_count(filename, /):
    try:
        with open(filename, 'r') as file:
            count = 0
            for line in file:
                stripped_line = line.strip() 
                if stripped_line and not stripped_line.startswith('#'):
                    count += 1
            return count
    except FileNotFoundError as e:
        raise FileNotFoundError("No such file") from e

# Quaternion Class
@dataclass(frozen=True)
class Quaternion:
    a: float
    b: float
    c: float
    d: float
    
    def __add__(self, other):
        if isinstance(other, Quaternion):
            return Quaternion(self.a + other.a, self.b + other.b, self.c + other.c, self.d + other.d)
        return NotImplemented
    
    def __mul__(self, other):
        if isinstance(other, Quaternion):
            a = self.a * other.a - self.b * other.b - self.c * other.c - self.d * other.d
            b = self.a * other.b + self.b * other.a + self.c * other.d - self.d * other.c
            c = self.a * other.c - self.b * other.d + self.c * other.a + self.d * other.b
            d = self.a * other.d + self.b * other.c - self.c * other.b + self.d * other.a
            return Quaternion(a, b, c, d)
        return NotImplemented
    
    def __eq__(self, other):
        if isinstance(other, Quaternion):
            return (self.a, self.b, self.c, self.d) == (other.a, other.b, other.c, other.d)
        return False
    
    @property
    def conjugate(self):
        return Quaternion(self.a, -self.b, -self.c, -self.d)
    
    @property
    def coefficients(self):
        return (self.a, self.b, self.c, self.d)
    
    def __str__(self):
        components = []
        
        if self.a != 0 or (self.a == 0 and self.b == 0 and self.c == 0 and self.d == 0):
            components.append(f"{self.a}")
        
        if self.b != 0:
            components.append(f"{'+' if self.b > 0 and components else ''}{'-' if self.b < 0 else ''}{'' if abs(self.b) == 1 else abs(self.b)}i")
        
        if self.c != 0:
            components.append(f"{'+' if self.c > 0 and components else ''}{'-' if self.c < 0 else ''}{'' if abs(self.c) == 1 else abs(self.c)}j")
        
        if self.d != 0:
            components.append(f"{'+' if self.d > 0 and components else ''}{'-' if self.d < 0 else ''}{'' if abs(self.d) == 1 else abs(self.d)}k")
        
        return "".join(components)